### DISCUSSION TITLE
Using 256 bit AES instead of 128 bit for TLS

### DISCUSSION THREAD

**Discussion by Grant12311** [nginx-proxy/nginx-proxy#2149] [Q&A]
Hello. I was checking the TLS details of the pages I'm now hosting behind this proxy and I noticed that their security has dropped somewhat from before. This proxy seems to use 128 bit AES instead of 256 bit like Nextcloud and at least some of my other services used on their own. I'd like to fix that but I can't seem to find a way to do it. I'm assuming Nginx supports 256 bit, though I'll admit I'm not sure and couldn't manage to find a straight answer online. Any help is greatly appreciated.

**Comment by buchdag** [comment]
Hi. If your are using the default Mozilla-Intermediate or the Mozilla-Modern `SSL_POLICY`, nginx is configured with [`ssl_prefer_server_ciphers off;`](http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_prefer_server_ciphers), which means the actual cipher suite used will be the one preferred by the client you're using.

You should be able to configure an overriding SSL policy using the existing [per-VIRTUAL_HOST mechanism](https://github.com/nginx-proxy/nginx-proxy#per-virtual_host).

**Comment by Grant12311** [comment]
Hi! Thanks for the reply. I tried my best to based on the information you provided but I couldn't get it to work. I "mounted" the /etc/nginx/vhost.d/ folder into a docker volume and made a file called "default" in it. Then I placed this in the file 

```
ssl_ciphers` AES256-SHA;
```

Nginx failed when I tried placing that in what is called a "server directive", so clearly I'm actually interacting with the server, just not correctly Please keep in mind that I've never used Nginx before today and never used Docker before yesterday so I expect I'm making a dumb mistake somewhere that I wouldn't if I was familiar with both/either.

**Comment by buchdag** [comment]
Hi @Grant12311 

The [ciphers suites](https://en.wikipedia.org/wiki/Cipher_suite) actually have more complexe names, here are those for the default SSL Policy:
- ECDHE-ECDSA-AES128-GCM-SHA256
- ECDHE-RSA-AES128-GCM-SHA256
- ECDHE-ECDSA-AES256-GCM-SHA384
- ECDHE-RSA-AES256-GCM-SHA384
- ECDHE-ECDSA-CHACHA20-POLY1305
- ECDHE-RSA-CHACHA20-POLY1305
- DHE-RSA-AES128-GCM-SHA256
- DHE-RSA-AES256-GCM-SHA384

`ECDHE` / `DHE` is the key exchange algorithm.
`ECDSA` / `RSA` is the signature algorithm.
`AES128-GCM` / `AES256-GCM` / `CHACHA20` is the bulk encryption algorithm.
`SHA256` / `SHA384` / `POLY1305` is the message authentication code (MAC) algorithm.

If you change your `/etc/nginx/vhost.d/ default` files to something like:
```nginx
ssl_ciphers 'ECDHE-ECDSA-AES256-GCM-SHA384:CDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384';
```

Nginx should only accept cipher suites that use `AES256-GCM` as a bulk encryption algorithm.

Please note however that AES128 for bulk encryption is considered  secure and that you won't really gain anything by forcing use of AES256:

> The difference between cracking the AES-128 algorithm and AES-256 algorithm is considered minimal. Whatever breakthrough might crack 128-bit will probably also crack 256-bit.

https://www.kryptall.com/index.php/information/how-safe-is-aes-encryption

My advice would be to keep the default configuration instead of restricting the set of cipher suites (for very hypothetical gains) and risking incompatibility with some clients.

> Please keep in mind that I've never used Nginx before today and never used Docker before yesterday so I expect I'm making a dumb mistake somewhere that I wouldn't if I was familiar with both/either.

That's what discussions are for 👍
